File: ts/packages/neutron-nucleus/src/features.ts
// ---------------------------------------------------------------------------
// Nucleus client — feature detection
// ---------------------------------------------------------------------------

import type { Transport, NucleusFeatures } from './types.js';

/**
 * Detect capabilities of the connected database.
 *
 * On a real Nucleus instance `SELECT VERSION()` will contain "Nucleus" and all
 * multi-model features are available. On plain PostgreSQL the result is a
 * standard PG version string and only SQL features are enabled.
 */
export async function detectFeatures(transport: Transport): Promise<NucleusFeatures> {
  const version = await transport.fetchval<string>('SELECT VERSION()');
  const ver = version ?? '';

  const isNucleus = ver.includes('Nucleus');

  return {
    isNucleus,
    hasKV: isNucleus,
    hasVector: isNucleus,
    hasTimeSeries: isNucleus,
    hasDocument: isNucleus,
    hasGraph: isNucleus,
    hasFTS: isNucleus,
    hasGeo: isNucleus,
    hasBlob: isNucleus,
    hasStreams: isNucleus,
    hasColumnar: isNucleus,
    hasDatalog: isNucleus,
    hasCDC: isNucleus,
    hasPubSub: isNucleus,
    version: ver,
  };
}

File: julia/src/features.jl
"""
    NucleusFeatures

Capabilities detected from VERSION() on connect. All fields are true when
connected to Nucleus, false when connected to plain PostgreSQL.
"""
struct NucleusFeatures
    is_nucleus::Bool
    version::Union{String, Nothing}
    has_kv::Bool
    has_vector::Bool
    has_timeseries::Bool
    has_document::Bool
    has_graph::Bool
    has_fts::Bool
    has_geo::Bool
    has_blob::Bool
    has_streams::Bool
    has_columnar::Bool
    has_datalog::Bool
    has_cdc::Bool
    has_pubsub::Bool
end

"""Parse VERSION() output and return NucleusFeatures."""
function detect_features(conn::LibPQ.Connection)::NucleusFeatures
    result = LibPQ.execute(conn, "SELECT VERSION()")
    version_str = first(result)[1]
    # Nucleus returns: "PostgreSQL 16.0 (Nucleus X.Y.Z — The Definitive Database)"
    is_nucleus = occursin("Nucleus", version_str)
    m = match(r"Nucleus (\S+)", version_str)
    nucleus_version = (is_nucleus && m !== nothing) ? m.captures[1] : nothing
    return NucleusFeatures(
        is_nucleus, nucleus_version,
        is_nucleus, is_nucleus, is_nucleus, is_nucleus,
        is_nucleus, is_nucleus, is_nucleus, is_nucleus,
        is_nucleus, is_nucleus, is_nucleus, is_nucleus,
        is_nucleus,
    )
end
